File: web/src/services/papers.ts
import { apiFetch } from './api'

export type PaperStatus = 'unread' | 'reading' | 'done'

export interface Paper {
  id: string
  rel_path: string
  title: string
  category: string
  size_bytes: number
  status: PaperStatus
  has_file: boolean
  started_on: string | null
  finished_on: string | null
}

export interface PaperGroup {
  category: string
  papers: Paper[]
}

export interface PaperStats {
  total_count: number
  unread_count: number
  reading_count: number
  done_count: number
}

export interface PaperListResponse {
  groups: PaperGroup[]
  stats: PaperStats
}

export interface PaperTodayResponse {
  date: string
  paper: Paper | null
  stats: PaperStats
}

export interface PaperStatusResult {
  paper: Paper
}

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const response = await apiFetch(`/api${path}`, {
    ...init,
    headers: {
      'Content-Type': 'application/json',
      ...init?.headers,
    },
  })
  if (!response.ok) {
    throw new Error(`Papers API 请求失败：${response.status}`)
  }
  return (await response.json()) as T
}

export function fetchPapers(): Promise<PaperListResponse> {
  return request<PaperListResponse>('/papers')
}

export function fetchPaperToday(): Promise<PaperTodayResponse> {
  return request<PaperTodayResponse>('/papers/today')
}

export function setPaperStatus(
  paperId: string,
  status: PaperStatus,
): Promise<PaperStatusResult> {
  return request<PaperStatusResult>(`/papers/${paperId}/status`, {
    method: 'POST',
    body: JSON.stringify({ status }),
  })
}

export async function openPaperFile(paperId: string): Promise<void> {
  const response = await apiFetch(`/api/papers/${paperId}/file`)
  if (!response.ok) {
    throw new Error(`PDF 加载失败：${response.status}`)
  }
  const blob = await response.blob()
  const url = URL.createObjectURL(blob)
  window.open(url, '_blank', 'noopener')
}

File: backend/src/graduate_entrance/schemas/papers.py
from __future__ import annotations

from datetime import date, datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field

PaperStatus = Literal["unread", "reading", "done"]
PaperContentSource = Literal["ar5iv", "pdf"]
AnnotationColor = Literal["yellow", "green", "blue", "red"]


class PaperSyncItem(BaseModel):
    rel_path: str = Field(min_length=1, max_length=500)
    title: str = ""
    category: str = ""
    size_bytes: int = Field(default=0, ge=0)


class PaperSyncRequest(BaseModel):
    papers: list[PaperSyncItem]


class PaperSyncResult(BaseModel):
    imported: int
    updated: int
    total_count: int


class PaperRead(BaseModel):
    id: UUID
    rel_path: str
    title: str
    category: str
    size_bytes: int
    status: PaperStatus
    has_file: bool
    has_content: bool = False
    started_on: date | None
    finished_on: date | None


class PaperGroup(BaseModel):
    category: str
    papers: list[PaperRead]


class PaperStatsResponse(BaseModel):
    total_count: int
    unread_count: int
    reading_count: int
    done_count: int


class PaperListResponse(BaseModel):
    groups: list[PaperGroup]
    stats: PaperStatsResponse


class PaperTodayResponse(BaseModel):
    date: date
    paper: PaperRead | None
    stats: PaperStatsResponse


class PaperStatusRequest(BaseModel):
    status: PaperStatus
    as_of: date | None = None


class PaperStatusResult(BaseModel):
    paper: PaperRead


class PaperBlock(BaseModel):
    type: Literal["heading", "paragraph"]
    md: str
    level: int = Field(default=0, ge=0, le=6)


class PaperContentUpload(BaseModel):
    source: PaperContentSource
    blocks: list[PaperBlock] = Field(min_length=1)


class PaperTocEntry(BaseModel):
    title: str
    level: int
    block_index: int


class PaperContentResponse(BaseModel):
    paper: PaperRead
    source: PaperContentSource
    blocks: list[PaperBlock]
    toc: list[PaperTocEntry]


class PaperAnnotationCreate(BaseModel):
    block_index: int = Field(ge=0)
    excerpt: str = ""
    note: str = ""
    color: AnnotationColor = "yellow"


class PaperAnnotationUpdate(BaseModel):
    note: str | None = None
    color: AnnotationColor | None = None


class PaperAnnotationRead(BaseModel):
    id: UUID
    paper_id: UUID
    block_index: int
    excerpt: str
    note: str
    color: AnnotationColor
    created_at: datetime


class PaperAnnotationList(BaseModel):
    annotations: list[PaperAnnotationRead]
